Colinux Block device


Index: linux-2.6.15-source/include/linux/major.h
===================================================================
--- linux-2.6.15-source.orig/include/linux/major.h
+++ linux-2.6.15-source/include/linux/major.h
@@ -130,6 +130,7 @@
 #define VIOCD_MAJOR		113
 
 #define ATARAID_MAJOR		114
+#define COLINUX_MAJOR		117
 
 #define SCSI_DISK8_MAJOR	128
 #define SCSI_DISK9_MAJOR	129
Index: linux-2.6.15-source/drivers/block/Kconfig
===================================================================
--- linux-2.6.15-source.orig/drivers/block/Kconfig
+++ linux-2.6.15-source/drivers/block/Kconfig
@@ -408,6 +408,15 @@
 	  "real" root file system, etc. See <file:Documentation/initrd.txt>
 	  for details.
 
+config BLK_DEV_COBD
+	tristate 'Cooperative block device support'
+	default y
+	depends on COOPERATIVE=y
+	help
+          Virtual block device support for cooperative kernels.
+
+          If unsure, say Y.
+
 
 config CDROM_PKTCDVD
 	tristate "Packet writing on CD/DVD media"
Index: linux-2.6.15-source/drivers/block/Makefile
===================================================================
--- linux-2.6.15-source.orig/drivers/block/Makefile
+++ linux-2.6.15-source/drivers/block/Makefile
@@ -15,6 +15,7 @@
 obj-$(CONFIG_ATARI_SLM)		+= acsi_slm.o
 obj-$(CONFIG_AMIGA_Z2RAM)	+= z2ram.o
 obj-$(CONFIG_BLK_DEV_RAM)	+= rd.o
+obj-$(CONFIG_BLK_DEV_COBD)	+= cobd.o
 obj-$(CONFIG_BLK_DEV_LOOP)	+= loop.o
 obj-$(CONFIG_BLK_DEV_PS2)	+= ps2esdi.o
 obj-$(CONFIG_BLK_DEV_XD)	+= xd.o
