Implements network support


Index: linux-2.6.22-source/drivers/net/Kconfig
===================================================================
--- linux-2.6.22-source.orig/drivers/net/Kconfig
+++ linux-2.6.22-source/drivers/net/Kconfig
@@ -119,6 +119,10 @@
 
 	  If you don't know what to use this for, you don't need it.
 
+config CONET_COOPERATIVE
+	tristate 'Cooperative Virtual Ethernet driver support'
+	depends on COOPERATIVE
+
 config NET_SB1000
 	tristate "General Instruments Surfboard 1000"
 	depends on PNP
@@ -152,7 +156,7 @@
 #
 
 menu "Ethernet (10 or 100Mbit)"
-	depends on !UML
+	depends on !UML && !COOPERATIVE
 
 config NET_ETHERNET
 	bool "Ethernet (10 or 100Mbit)"
@@ -1900,7 +1904,7 @@
 
 menuconfig NETDEV_1000
 	bool "Ethernet (1000 Mbit)"
-	depends on !UML
+	depends on !UML && !COOPERATIVE
 	default y
 
 if NETDEV_1000
@@ -2338,7 +2342,7 @@
 
 menuconfig NETDEV_10000
 	bool "Ethernet (10000 Mbit)"
-	depends on !UML
+	depends on !UML && !COOPERATIVE
 	default y
 
 if NETDEV_10000
Index: linux-2.6.22-source/drivers/net/Makefile
===================================================================
--- linux-2.6.22-source.orig/drivers/net/Makefile
+++ linux-2.6.22-source/drivers/net/Makefile
@@ -163,6 +163,7 @@
 
 # This is also a 82596 and should probably be merged
 obj-$(CONFIG_LP486E) += lp486e.o
+obj-$(CONFIG_CONET_COOPERATIVE) += conet.o
 
 obj-$(CONFIG_ETH16I) += eth16i.o
 obj-$(CONFIG_ZORRO8390) += zorro8390.o
