Various drivers


Index: linux-2.6.13-source/drivers/block/Kconfig
===================================================================
--- linux-2.6.13-source.orig/drivers/block/Kconfig
+++ linux-2.6.13-source/drivers/block/Kconfig
@@ -408,6 +408,15 @@
 	  "real" root file system, etc. See <file:Documentation/initrd.txt>
 	  for details.
 
+config BLK_DEV_COBD
+	tristate 'Cooperative block device support'
+	default y
+	depends on COOPERATIVE=y
+	help
+	  Virtual block device support for cooperative kernels.
+
+	  If unsure, say Y.
+
 config INITRAMFS_SOURCE
 	string "Initramfs source file(s)"
 	default ""
Index: linux-2.6.13-source/drivers/block/Makefile
===================================================================
--- linux-2.6.13-source.orig/drivers/block/Makefile
+++ linux-2.6.13-source/drivers/block/Makefile
@@ -29,6 +29,7 @@
 obj-$(CONFIG_ATARI_SLM)		+= acsi_slm.o
 obj-$(CONFIG_AMIGA_Z2RAM)	+= z2ram.o
 obj-$(CONFIG_BLK_DEV_RAM)	+= rd.o
+obj-$(CONFIG_BLK_DEV_COBD)	+= cobd.o
 obj-$(CONFIG_BLK_DEV_LOOP)	+= loop.o
 obj-$(CONFIG_BLK_DEV_PS2)	+= ps2esdi.o
 obj-$(CONFIG_BLK_DEV_XD)	+= xd.o
Index: linux-2.6.13-source/drivers/char/mem.c
===================================================================
--- linux-2.6.13-source.orig/drivers/char/mem.c
+++ linux-2.6.13-source/drivers/char/mem.c
@@ -25,6 +25,7 @@
 #include <linux/device.h>
 #include <linux/highmem.h>
 #include <linux/crash_dump.h>
+#include <linux/cooperative_internal.h>
 #include <linux/backing-dev.h>
 #include <linux/bootmem.h>
 
@@ -118,6 +119,11 @@
 	ssize_t read, sz;
 	char *ptr;
 
+	if (cooperative_mode_enabled()) {
+		/* Reading kmem is bad news in cooperative */
+		return -ENOMEM;
+	}
+
 	if (!valid_phys_addr_range(p, &count))
 		return -EFAULT;
 	read = 0;
@@ -176,6 +182,11 @@
 	unsigned long copied;
 	void *ptr;
 
+	if (cooperative_mode_enabled()) {
+		/* Writing kmem is bad news in cooperative */
+		return -ENOMEM;
+	}
+
 	if (!valid_phys_addr_range(p, &count))
 		return -EFAULT;
 
@@ -234,6 +245,9 @@
 
 static int mmap_mem(struct file * file, struct vm_area_struct * vma)
 {
+	if (cooperative_mode_enabled())
+		return -EFAULT;
+
 #if defined(__HAVE_PHYS_MEM_ACCESS_PROT)
 	unsigned long offset = vma->vm_pgoff << PAGE_SHIFT;
 
@@ -263,6 +277,9 @@
 {
 	unsigned long pfn;
 
+	if (cooperative_mode_enabled())
+		return -EFAULT;
+
 	/* Turn a kernel-virtual address into a physical page frame */
 	pfn = __pa((u64)vma->vm_pgoff << PAGE_SHIFT) >> PAGE_SHIFT;
 
@@ -327,6 +344,11 @@
 	ssize_t low_count, read, sz;
 	char * kbuf; /* k-addr because vread() takes vmlist_lock rwlock */
 
+	if (cooperative_mode_enabled()) {
+		/* Reading kmem is bad news in cooperative */
+		return -ENOMEM;
+	}
+
 	read = 0;
 	if (p < (unsigned long) high_memory) {
 		low_count = count;
@@ -478,6 +500,11 @@
 	ssize_t written;
 	char * kbuf; /* k-addr because vwrite() takes vmlist_lock rwlock */
 
+	if (cooperative_mode_enabled()) {
+		/* Writing kmem is bad news in cooperative */
+		return -ENOMEM;
+	}
+
 	if (p < (unsigned long) high_memory) {
 
 		wrote = count;
Index: linux-2.6.13-source/drivers/input/keyboard/Kconfig
===================================================================
--- linux-2.6.13-source.orig/drivers/input/keyboard/Kconfig
+++ linux-2.6.13-source/drivers/input/keyboard/Kconfig
@@ -16,8 +16,8 @@
 	tristate "AT keyboard" if !PC
 	default y
 	select SERIO
-	select SERIO_LIBPS2
-	select SERIO_I8042 if PC
+ 	select SERIO_LIBPS2
+	select SERIO_I8042 if PC && !COOPERATIVE
 	select SERIO_GSCPS2 if GSC
 	help
 	  Say Y here if you want to use a standard AT or PS/2 keyboard. Usually
Index: linux-2.6.13-source/drivers/input/keyboard/atkbd.c
===================================================================
--- linux-2.6.13-source.orig/drivers/input/keyboard/atkbd.c
+++ linux-2.6.13-source/drivers/input/keyboard/atkbd.c
@@ -27,6 +27,7 @@
 #include <linux/serio.h>
 #include <linux/workqueue.h>
 #include <linux/libps2.h>
+#include <linux/cooperative_internal.h>
 
 #define DRIVER_DESC	"AT and PS/2 keyboard driver"
 
@@ -513,6 +514,9 @@
 	struct ps2dev *ps2dev = &atkbd->ps2dev;
 	unsigned char param[2];
 
+	if (cooperative_mode_enabled())
+		return 0;
+
 /*
  * Some systems, where the bit-twiddling when testing the io-lines of the
  * controller may confuse the keyboard need a full reset of the keyboard. On
Index: linux-2.6.13-source/drivers/input/mouse/Kconfig
===================================================================
--- linux-2.6.13-source.orig/drivers/input/mouse/Kconfig
+++ linux-2.6.13-source/drivers/input/mouse/Kconfig
@@ -15,6 +15,7 @@
 config MOUSE_PS2
 	tristate "PS/2 mouse"
 	default y
+	depends on !COOPERATIVE
 	select SERIO
 	select SERIO_LIBPS2
 	select SERIO_I8042 if PC
@@ -37,8 +38,21 @@
 	  To compile this driver as a module, choose M here: the
 	  module will be called psmouse.
 
+config MOUSE_COOPERATIVE
+	tristate "Cooperative Mouse driver"
+       	default y
+        depends on COOPERATIVE
+        ---help---
+          Virtual mouse driver for cooperative kernels.
+
+          If unsure, say Y.
+
+          To compile this driver as a module, choose M here: the
+          module will be called psmouse.
+
 config MOUSE_SERIAL
 	tristate "Serial mouse"
+	depends on !COOPERATIVE
 	select SERIO
 	---help---
 	  Say Y here if you have a serial (RS-232, COM port) mouse connected
@@ -52,7 +66,7 @@
 
 config MOUSE_INPORT
 	tristate "InPort/MS/ATIXL busmouse"
-	depends on ISA
+	depends on ISA && !COOPERATIVE
 	help
 	  Say Y here if you have an InPort, Microsoft or ATI XL busmouse.
 	  They are rather rare these days.
@@ -62,13 +76,13 @@
 
 config MOUSE_ATIXL
 	bool "ATI XL variant"
-	depends on MOUSE_INPORT
+	depends on MOUSE_INPORT && !COOPERATIVE
 	help
 	  Say Y here if your mouse is of the ATI XL variety.
 
 config MOUSE_LOGIBM
 	tristate "Logitech busmouse"
-	depends on ISA
+	depends on ISA && !COOPERATIVE
 	help
 	  Say Y here if you have a Logitech busmouse.
 	  They are rather rare these days.
@@ -78,7 +92,7 @@
 
 config MOUSE_PC110PAD
 	tristate "IBM PC110 touchpad"
-	depends on ISA
+	depends on ISA && !COOPERATIVE
 	help
 	  Say Y if you have the IBM PC-110 micro-notebook and want its
 	  touchpad supported.
@@ -88,7 +102,7 @@
 
 config MOUSE_MAPLE
 	tristate "Maple bus mouse"
-	depends on SH_DREAMCAST && MAPLE
+	depends on SH_DREAMCAST && MAPLE && !COOPERATIVE
 	help
 	  Say Y if you have a DreamCast console and a mouse attached to
 	  its Maple bus.
@@ -98,7 +112,7 @@
 
 config MOUSE_AMIGA
 	tristate "Amiga mouse"
-	depends on AMIGA
+	depends on AMIGA && !COOPERATIVE
 	help
 	  Say Y here if you have an Amiga and want its native mouse
 	  supported by the kernel.
@@ -108,7 +122,7 @@
 
 config MOUSE_RISCPC
 	tristate "Acorn RiscPC mouse"
-	depends on ARCH_ACORN
+	depends on ARCH_ACORN && !COOPERATIVE
 	help
 	  Say Y here if you have the Acorn RiscPC computer and want its
 	  native mouse supported.
@@ -118,6 +132,7 @@
 
 config MOUSE_VSXXXAA
 	tristate "DEC VSXXX-AA/GA mouse and VSXXX-AB tablet"
+	depends on !COOPERATIVE
 	select SERIO
 	help
 	  Say Y (or M) if you want to use a DEC VSXXX-AA (hockey
@@ -129,10 +144,9 @@
 
 config MOUSE_HIL
 	tristate "HIL pointers (mice etc)."     
-	depends on GSC
+	depends on GSC && !COOPERATIVE
 	select HP_SDC
 	select HIL_MLC
 	help
 	  Say Y here to support HIL pointers.
-
 endif
Index: linux-2.6.13-source/drivers/input/mouse/Makefile
===================================================================
--- linux-2.6.13-source.orig/drivers/input/mouse/Makefile
+++ linux-2.6.13-source/drivers/input/mouse/Makefile
@@ -11,6 +11,7 @@
 obj-$(CONFIG_MOUSE_MAPLE)	+= maplemouse.o
 obj-$(CONFIG_MOUSE_PC110PAD)	+= pc110pad.o
 obj-$(CONFIG_MOUSE_PS2)		+= psmouse.o
+obj-$(CONFIG_MOUSE_COOPERATIVE)	+= comouse.o
 obj-$(CONFIG_MOUSE_SERIAL)	+= sermouse.o
 obj-$(CONFIG_MOUSE_HIL)		+= hil_ptr.o
 obj-$(CONFIG_MOUSE_VSXXXAA)	+= vsxxxaa.o
Index: linux-2.6.13-source/drivers/input/mouse/comouse.c
===================================================================
--- /dev/null
+++ linux-2.6.13-source/drivers/input/mouse/comouse.c
@@ -0,0 +1,74 @@
+/*
+ * Virtual mouse driver for Linux
+ *
+ * Skeleton based on:
+ *  $Id: comouse.c,v 1.1.6.1 2005/07/15 08:19:48 tamura Exp $
+ *
+ *  Copyright (c) 1999-2001 Vojtech Pavlik
+ *
+ * Copyright (c) 2004 Dan Aloni
+ */
+
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/interrupt.h>
+#include <linux/input.h>
+#include <linux/config.h>
+#include <linux/serio.h>
+#include <linux/init.h>
+
+MODULE_AUTHOR("Dan Aloni <da-x@colinux.org>");
+MODULE_DESCRIPTION("Virtual mouse driver");
+MODULE_LICENSE("GPL");
+
+/*
+ * comouse_interrupt() handles incoming characters, either gathering them into
+ * packets or passing them to the command routine as command output.
+ */
+
+static irqreturn_t comouse_interrupt(struct serio *serio,
+				     unsigned char data, unsigned int flags, struct pt_regs *regs)
+{
+	return IRQ_HANDLED;
+}
+
+/*
+ * comouse_disconnect() cleans up after we don't want talk
+ * to the mouse anymore.
+ */
+
+static void comouse_disconnect(struct serio *serio)
+{
+}
+
+/*
+ * comouse_connect() is a callback form the serio module when
+ * an unhandled serio port is found.
+ */
+
+static void comouse_connect(struct serio *serio, struct serio_driver *dev)
+{
+}
+
+static struct serio_driver comouse_dev = {
+	.interrupt =	comouse_interrupt,
+	.connect =	comouse_connect,
+	.disconnect =	comouse_disconnect,
+	.driver		= {
+		.name	= "comouse",
+	},
+};
+
+int __init comouse_init(void)
+{
+	serio_register_driver(&comouse_dev);
+	return 0;
+}
+
+void __exit comouse_exit(void)
+{
+	serio_unregister_driver(&comouse_dev);
+}
+
+module_init(comouse_init);
+module_exit(comouse_exit);
Index: linux-2.6.13-source/drivers/input/serio/Kconfig
===================================================================
--- linux-2.6.13-source.orig/drivers/input/serio/Kconfig
+++ linux-2.6.13-source/drivers/input/serio/Kconfig
@@ -21,7 +21,7 @@
 config SERIO_I8042
 	tristate "i8042 PC Keyboard controller" if EMBEDDED || !X86
 	default y
-	depends on !PARISC && (!ARM || ARCH_SHARK || FOOTBRIDGE_HOST) && !M68K
+	depends on !PARISC && (!ARM || ARCH_SHARK || FOOTBRIDGE_HOST) && !M68K && !COOPERATIVE
 	---help---
 	  i8042 is the chip over which the standard AT keyboard and PS/2
 	  mouse are connected to the computer. If you use these devices,
@@ -180,4 +180,9 @@
 	  To compile this driver as a module, choose M here: the
 	  module will be called serio_raw.
 
+config SERIO_COKBD
+ 	tristate "Cooperative Linux virtual keyboard controller driver"
+ 	depends on COOPERATIVE
+ 	default y
+
 endif
Index: linux-2.6.13-source/drivers/input/serio/Makefile
===================================================================
--- linux-2.6.13-source.orig/drivers/input/serio/Makefile
+++ linux-2.6.13-source/drivers/input/serio/Makefile
@@ -21,3 +21,4 @@
 obj-$(CONFIG_SERIO_MACEPS2)	+= maceps2.o
 obj-$(CONFIG_SERIO_LIBPS2)	+= libps2.o
 obj-$(CONFIG_SERIO_RAW)		+= serio_raw.o
+obj-$(CONFIG_SERIO_COKBD)	+= cokbd.o
\ No newline at end of file
