--- fltk-1.1.4/src/Fl_win32.cxx	2003-06-15 07:27:35.000000000 +0300
+++ fltk-1.1.4/src/Fl_win32.cxx	2004-04-10 22:06:11.000000000 +0300
@@ -47,7 +47,6 @@
 // The following include files require GCC 3.x or a non-GNU compiler...
 #if !defined(__GNUC__) || __GNUC__ >= 3
 #  include <ole2.h>
-#  include <ShellApi.h>
 #endif // !__GNUC__ || __GNUC__ >= 3
 
 
diff -urN fltk-1.1.4/configure fltk-1.1.4-patch-noX/configure
--- fltk-1.1.4-patch-noX/configure
+++ fltk-1.1.4/configure
@@ -8040,7 +8040,7 @@
 fi
 
 
-	if test x$no_x = xyes; then
+	if test x$no_x = xyes -a x$have_x != xdisabled; then
 	    { { echo "$as_me:$LINENO: error: Configure could not find required X11 libraries" >&5
 echo "$as_me: error: Configure could not find required X11 libraries" >&2;}
    { (exit aborting.); exit aborting.; }; }
