Serial support

Index: linux-2.6.26-source/drivers/char/Makefile
===================================================================
--- linux-2.6.26-source.orig/drivers/char/Makefile
+++ linux-2.6.26-source/drivers/char/Makefile
@@ -26,6 +26,7 @@
 obj-$(CONFIG_CYCLADES)		+= cyclades.o
 obj-$(CONFIG_STALLION)		+= stallion.o
 obj-$(CONFIG_ISTALLION)		+= istallion.o
+obj-$(CONFIG_SERIAL_COOPERATIVE) += coserial.o
 obj-$(CONFIG_NOZOMI)		+= nozomi.o
 obj-$(CONFIG_DIGIEPCA)		+= epca.o
 obj-$(CONFIG_SPECIALIX)		+= specialix.o
Index: linux-2.6.26-source/drivers/char/Kconfig
===================================================================
--- linux-2.6.26-source.orig/drivers/char/Kconfig
+++ linux-2.6.26-source/drivers/char/Kconfig
@@ -420,6 +420,13 @@
          If you have an SGI Altix with an attached SABrick
          say Y or M here, otherwise say N.
 
+config SERIAL_COOPERATIVE
+       bool "coLinux serial device support"
+       depends on COOPERATIVE
+       default y
+       help
+         say Y, if you have colinux
+
 source "drivers/serial/Kconfig"
 
 config UNIX98_PTYS
