Serial support


Index: linux-2.6.14-source/drivers/char/Makefile
===================================================================
--- linux-2.6.14-source.orig/drivers/char/Makefile
+++ linux-2.6.14-source/drivers/char/Makefile
@@ -25,6 +25,7 @@
 obj-$(CONFIG_CYCLADES)		+= cyclades.o
 obj-$(CONFIG_STALLION)		+= stallion.o
 obj-$(CONFIG_ISTALLION)		+= istallion.o
+obj-$(CONFIG_COOPERATIVE)     	+= cocd.o
 obj-$(CONFIG_DIGIEPCA)		+= epca.o
 obj-$(CONFIG_SPECIALIX)		+= specialix.o
 obj-$(CONFIG_MOXA_INTELLIO)	+= moxa.o
