Index: linux-2.6.33-source/drivers/video/Kconfig
===================================================================
--- linux-2.6.33-source.orig/drivers/video/Kconfig
+++ linux-2.6.33-source/drivers/video/Kconfig
@@ -242,6 +242,23 @@
 comment "Frame buffer hardware drivers"
 	depends on FB
 
+config COOPERATIVE_VIDEO
+	bool "Cooperative Video"
+	default n
+	depends on COOPERATIVE
+	select FB
+        select FB_SYS_FILLRECT
+        select FB_SYS_COPYAREA
+        select FB_SYS_IMAGEBLIT
+        select FB_SYS_FOPS
+	select FRAMEBUFFER_CONSOLE if !COOPERATIVE_CONSOLE
+	select LOGO if !COOPERATIVE_CONSOLE
+	select LOGO_LINUX_CLUT224 if !COOPERATIVE_CONSOLE
+	---help---
+	  This enables the Cooperative Framebuffer Video driver.
+
+	  Say N unless you are compiling a Cooperative kernel.
+
 config FB_CIRRUS
 	tristate "Cirrus Logic support"
 	depends on FB && (ZORRO || PCI)
@@ -2062,6 +2079,17 @@
 	  found on DA8xx/OMAP-L1xx SoCs.
 	  If unsure, say N.
 
+config FB_COFB
+        bool "Cooperative Virtual Frame Buffer support"
+       depends on FB && COOPERATIVE
+       ---help---
+         This is a `virtual' frame buffer device for coLinux.
+
+         The driver cannot (currently) be built as a module.
+
+         If unsure, say Y.
+
+
 config FB_VIRTUAL
 	tristate "Virtual Frame Buffer support (ONLY FOR TESTING!)"
 	depends on FB
Index: linux-2.6.33-source/drivers/video/Makefile
===================================================================
--- linux-2.6.33-source.orig/drivers/video/Makefile
+++ linux-2.6.33-source/drivers/video/Makefile
@@ -137,6 +137,7 @@
 obj-$(CONFIG_FB_EFI)              += efifb.o
 obj-$(CONFIG_FB_VGA16)            += vga16fb.o
 obj-$(CONFIG_FB_OF)               += offb.o
+obj-$(CONFIG_FB_COFB)             += covideo.o cfbfillrect.o cfbcopyarea.o cfbimgblt.o
 obj-$(CONFIG_FB_BF54X_LQ043)	  += bf54x-lq043fb.o
 obj-$(CONFIG_FB_BFIN_LQ035Q1)     += bfin-lq035q1-fb.o
 obj-$(CONFIG_FB_BFIN_T350MCQB)	  += bfin-t350mcqb-fb.o
Index: linux-2.6.33-source/drivers/gpu/vga/Kconfig
===================================================================
--- linux-2.6.33-source.orig/drivers/gpu/vga/Kconfig
+++ linux-2.6.33-source/drivers/gpu/vga/Kconfig
@@ -1,7 +1,7 @@
 config VGA_ARB
 	bool "VGA Arbitration" if EMBEDDED
 	default y
-	depends on PCI
+	depends on PCI && !COOPERATIVE
 	help
 	  Some "legacy" VGA devices implemented on PCI typically have the same
 	  hard-decoded addresses as they did on ISA. When multiple PCI devices
Index: linux-2.6.33-source/mm/memory.c
===================================================================
--- linux-2.6.33-source.orig/mm/memory.c
+++ linux-2.6.33-source/mm/memory.c
@@ -1693,7 +1693,11 @@
 	arch_enter_lazy_mmu_mode();
 	do {
 		BUG_ON(!pte_none(*pte));
+#ifndef CONFIG_COOPERATIVE
 		set_pte_at(mm, addr, pte, pte_mkspecial(pfn_pte(pfn, prot)));
+#else
+                set_pte_at(mm, addr, pte, pte_mkspecial(__pte((pfn<<PAGE_SHIFT)|pgprot_val(prot))));
+#endif
 		pfn++;
 	} while (pte++, addr += PAGE_SIZE, addr != end);
 	arch_leave_lazy_mmu_mode();
